src/sgp/SoundMan.cc

```diff
@@ -467,6 +463,111 @@ void SoundStopAllRandom(void)
+			if (result != MA_SUCCESS) {
+				throw std::runtime_error(ST::format("ma_decoder_seek_to_pcm_frame: {}", ma_result_description(result)).c_str());
+			}
+			// We stream from file
+			framesRead = ma_decoder_read_pcm_frames(sample->pDecoder, pFramesInClientFormat, bytesToWrite);
+		} else if (sample->pDataConverter != NULL) {
+			auto bytesPerFrame = ma_get_bytes_per_frame(sample->eInMemoryFormat, sample->uiInMemoryChannels);
+			auto posInBytes = ma_data_converter_get_required_input_frame_count(sample->pDataConverter, channel->Pos) * bytesPerFrame;
+			auto requiredInputFrameCount = ma_data_converter_get_required_input_frame_count(sample->pDataConverter, bytesToWrite);
+			// We might not have as many bytes available
+			auto availableFrames = MIN(requiredInputFrameCount * bytesPerFrame, sample->uiBufferSize - posInBytes) / bytesPerFrame;
+			auto expectedOutputFrameCount = ma_data_converter_get_expected_output_frame_count(sample->pDataConverter, availableFrames);
+			
+			auto result = ma_data_converter_process_pcm_frames(
+				sample->pDataConverter,
+				sample->pInMemoryBuffer + posInBytes,
+				&availableFrames,
+				pFramesInClientFormat,
+				&expectedOutputFrameCount
+			);
+			if (result != MA_SUCCESS) {
+				throw std::runtime_error(ST::format("ma_data_converter_process_pcm_frames: {}", ma_result_description(result)).c_str());
+			}
+			framesRead = expectedOutputFrameCount;
+		} else {
+			throw std::runtime_error("Dont know how to process ring buffer");
+		}
+		result = ma_pcm_rb_commit_write(channel->pRingBuffer, framesRead, pFramesInClientFormat);
+		if (result != MA_SUCCESS) {
+			throw std::runtime_error(ST::format("ma_pcm_rb_commit_write: {}", ma_result_description(result)).c_str());
+		}
+
+		channel->Pos += framesRead;
+		if (framesRead < bytesToWrite) {
+			// If the sound is looped, continue to fill buffer in the next iteration
+			if (channel->Loops > 1) {
+				channel->Loops -= 1;
+				channel->Pos = 0;
+			} else {
+				channel->DoneServicing = TRUE;
+			}
+		}
+	} catch (const std::runtime_error& err) {
+		SLOGE(ST::format(
+			"Error processing audio stream for channel {}, sample {}, file \"{}\": {}",
+			channel - pSoundList, sample - pSampleList, sample->pName,
+			err.what()
+		).c_str());
+	}
+}
+
+static int SoundServiceBuffers(void *_ptr)
+{
+	SLOGD("Started SoundManBufferServiceThread");
+	while (1) {
+		std::unique_lock<std::mutex> lk(mutexBuffersNeedService);
+		conditionBuffersNeedService.wait(lk, []{
+			return fBuffersNeedService || fShutdownBufferServiceThread;
+		});
+		if (fShutdownBufferServiceThread) {
+			SLOGD("Stopped SoundManBufferServiceThread");
+			return 0;
+		}
+		if (fBuffersNeedService) {
+			for (UINT32 i = 0; i < lengthof(pSoundList); i++)
+			{
+				SOUNDTAG* Sound = &pSoundList[i];
+				if (Sound->State == CHANNEL_PLAY) {
+					FillRingBuffer(Sound);
+				}
+			}
+			fBuffersNeedService = FALSE;
+		}
+		lk.unlock();
```

Comment:
Do we know, in practice, how long is this critical section?  If I am understanding correctly, the main thread will wait on this mutex to update the `fBuffersNeedService` flag. 

My next question is - do we really need a strict mutex lock here? For example, is a simpler `std::atomic` flag enough? Or can we release the lock earlier? The worst case is that the main thread will update the `fBuffersNeedService` during the servicing, and that will trigger another servicing right away.

Reply:
Small correction: The main thread (where the game logic runs) is not involved in this anymore. It is only the `SoundManBufferServiceThread` and the audio callback, which afaik also runs in a separate thread (but I actually cannot find the documentation for this fact anymore :slightly_frowning_face:).

I think you are right with the assumption that, if this section takes too long, we block the audio callback thread, which is bad an can result in distorted audio. And the runtime depends quite a lot on the hardware used, as e.g. file reading is involved. We could use `mutex.try_lock` in the sound callback in order to mitigate this issue, because if the mutex is locked already, that just means we already are servicing the buffers and dont need to change `fBuffersNeedService` anyways.

Regarding whether we need the mutex: I think it is currently only required for the `condition_variable`, so we can make the `SoundManBufferServiceThread` wait (without using CPU resources) until there is actually something to service. I am not sure how to achieve this using other means, but if you have suggestions on how to make the thread wait without `std::condition_variable`, let me know...

Reply:
I tried digging into the documentation and Google, and came to the same conclusion as yours. It looks like `std::condition_varable` is the only way to do this (up to C++ 14), and thus a mutex must be involved. 

Then I think the current implementation is good as long as  the critical section is made as short as possible.

There is actually no multi-thread locking and synchronization. It may be a use-case of `std::async`, so that we make an async future to service the buffer whenever buffer is low.

Reply:
I changed the sound callback to use `try_lock`, so the length of the critical section should not matter anymore...